### DISCUSSION TITLE
Convergence problem encountered when calculating contact

### DISCUSSION THREAD

**Discussion by jin0123456** [idaholab/moose#27303] [Q&A Modules: General] [answered]
Hello,

I'm simulating a thermodynamic coupling problem, and my geometry model is a two cylindrical geometry with a gap. I've set the grid in RZ coordinates for ease of calculation. In the process of calculation, the gap would go from no contact to contact, and in the process I ran into convergence problems.
I set the penalty in the contact module `[gap_contact]` and output volume of the gap. If I set penalty large it would be not converged. On the contrary the volume of the gap becomes negative which I think the reason is the penalty factor too small the inner cylinder seeps into the outside. At last I set penalty 6.7e7 the volume would not be negative but there is a convergence problem. It would shorten the time step until it crashed. I upload the input/output file and mesh there.
https://github.com/jin0123456/miniature-system
Hope someone can give me some advice. Much appreciated.

**Comment by GiudGiud** [comment, marked as answer]
Hello

i think this is a known issue with penalty based approaches. If the penalty is too small, penetration can happen, if it s too large, the conditioning of the matrix takes a hit and it becomes difficult to solve the equations.

I think moose has Lagrange-multiplier and mortar based approaches to contact. You could try those?

Guillajme

**Reply by jin0123456** [reply]
Thanks for your advice. I have solved this problem by using mortar based approach.
By the way, can this approach be used on 3D geometry? Is the effect good?

**Reply by GiudGiud** [reply]
Hello
We have 3D mortar as of 18 months ago. Please feel free to try it

**Reply by jin0123456** [reply]
I see. Much appreciated.
